### PR TITLE
misc bugs

### PR DESCRIPTION
**Pull request by jschueller**
for the EDF service contract with @josephmure @mbaudin47 

- 1. IntervalMesher: Fix diamond=True in dim=3
#1670
The main problem is the simplices loop tried to add the 6 faces centers creating duplicates:
one solution is to add the 3 face centers on the lower bound separately first, then for each cell we add only the 3 face centers of the opposite sides plus the middle vertex
diamond

```
import openturns as ot
for diamond in [False, True]:
    mesher3D = ot.IntervalMesher([2,3,2])
    mesh3D = mesher3D.build(ot.Interval(3), diamond)
    mesh3D.exportToVTKFile(f"/tmp/mesh{diamond}.vtk")
```

- 2. Devised guidelines to implement object interface comparison, see inside #1245
and now `PersistentObject::operator==` compares the pointer instead answering True inconditionally
this will return a result just a dumb, but now the comparison will be correct between unrelated classes (#1896)

- 3. Improved PosteriorDistribution for #1552 (something is up with the CDF computation & normalization factor)

### CODE REVIEW

**Comment by jschueller** [comment]
TODO: move to experimental

**Review by regislebrun — APPROVED** [review]
PosteriorDistribution needs a full rework, otherwise good to go

File: python/src/ConditionalDistribution_doc.i.in [hunk 1]
```diff
@@ -85,7 +78,7 @@ Draw a sample:
 // ---------------------------------------------------------------------
 
 %feature("docstring") OT::ConditionalDistribution::getConditionedDistribution
```
Comment (josephmure) [inline python/src/ConditionalDistribution_doc.i.in:80]: The formatting is weird: why are the returns bold?

image

File: python/src/ConditionalDistribution_doc.i.in [hunk 2]
```diff
@@ -107,19 +100,19 @@ conditioningDistribution : :class:`~openturns.Distribution`, the distribution of
 // ---------------------------------------------------------------------
 
 %feature("docstring") OT::ConditionalDistribution::getLinkFunction
-"Accessor to the distribution's conditioned distribution parameter `linkFunction`.
+"Accessor to the distribution's conditioned distribution parameter *linkFunction*.
 
 Returns
 -------
 linkFunction : :class:`~openturns.Function`, the function
       :math:`g: \Rset^p \rightarrow \Rset^q`, where :math:`p` is the dimension of
```
Comment (josephmure) [inline python/src/ConditionalDistribution_doc.i.in:108]:
```suggestion
linkFunction : :class:`~openturns.Function`
      the function :math:`g: \Rset^p \rightarrow \Rset^q`, where :math:`p` is the dimension of
```

File: python/src/ConditionalDistribution_doc.i.in [hunk 3]
```diff
@@ -107,19 +100,19 @@ conditioningDistribution : :class:`~openturns.Distribution`, the distribution of
...
-      `conditioningDist` and :math:`q` is the dimension of :math:`\vect{\Theta}`.
+      *conditioningDist* and :math:`q` is the dimension of :math:`\vect{\Theta}`.
 "
 
 // ---------------------------------------------------------------------
 
 %feature("docstring") OT::ConditionalDistribution::setConditionedDistribution
```
Comment (josephmure) [inline python/src/ConditionalDistribution_doc.i.in:114]: Same formatting error

image

File: python/src/PosteriorDistribution_doc.i.in [hunk 4]
```diff
@@ -1,23 +1,33 @@
 %feature("docstring") OT::PosteriorDistribution
```
Comment (josephmure) [inline python/src/PosteriorDistribution_doc.i.in:1]: The experimental warning is missing.

File: python/src/PosteriorDistribution_doc.i.in [hunk 5]
```diff
@@ -1,23 +1,33 @@
...
 "Distribution conditioned by observations.
 
-Available constructors:
-    PosteriorDistribution(*conditional, observations*)
-
 Parameters
 ----------
 conditional : :class:`~openturns.ConditionalDistribution`
-    The conditional distribution.
+    The conditional distribution of density :math:`f_{\vect{X}}(\vect{x}) = \int f_{\vect{X}|\vect{\Theta}=\vect{y}}(\vect{x}|\vect{y}) f_{\vect{Y}}(\vect{y})\di{\vect{y}}`.
```
Comment (josephmure) [inline python/src/PosteriorDistribution_doc.i.in:7]:
```suggestion
    The distribution of density :math:`f_{\vect{X}}(\vect{x}) = \int f_{\vect{X}|\vect{\Theta}=\vect{y}}(\vect{x}|\vect{y}) f_{\vect{Y}}(\vect{y})\di{\vect{y}}`.
```
I know the class is named that way but the name is badly chosen, so let's try to keep the confusion to a minimum.

File: python/src/PosteriorDistribution_doc.i.in [hunk 6]
```diff
@@ -1,23 +1,33 @@
...
-ConditionalDistribution, MetropolisHastingsImplementation
+ConditionalDistribution, MetropolisHastings
+
+Notes
+-----
+Its probability density function is defined as:
+
+.. math::
+
+    f_{\vect{Y}|\vect{X}=\vect{x}}(\vect{y}) = f_{\vect{Y}}(\vect{y})\prod_{i=1}^n f_{X|\vect{\Theta}=\vect{y}}(x_i)\frac{1}{\prod_{i=1}^n f_{X}(x_i)}
```
Comment (josephmure) [inline python/src/PosteriorDistribution_doc.i.in:21]:
```suggestion
    f_{\vect{Y}|\vect{X}=\vect{x}}(\vect{y}) = \frac{f_{\vect{Y}}(\vect{y})\prod_{i=1}^n f_{X|\vect{\Theta}=\vect{y}}(x_i)}{\int f_{\vect{Y}}(\vect{y}')\prod_{i=1}^n f_{X|\vect{\Theta}=\vect{y}'}(x_i) d \vect{y}'}
```
Sorry for the confusion, but this is the correct formula.

**Review by josephmure — CHANGES_REQUESTED** [review]

**Comment by jschueller** [comment]
can I merge this soon @josephmure ? else it will go to 1.23

**Review by josephmure — APPROVED** [review]

**Comment by josephmure** [comment]
Yes, by all means, do merge it! @mbaudin47 will not have time to review this PR.

**Pull request merged**